[CI] CartesianShapeQueryTests testQueryRandomGeoCollection failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/p7n7etkicebme/tests/:x-pack:plugin:spatial:internalClusterTest/org.elasticsearch.xpack.spatial.search.CartesianShapeQueryTests/testQueryRandomGeoCollection

**Reproduction line:**
`./gradlew ':x-pack:plugin:spatial:internalClusterTest' --tests "org.elasticsearch.xpack.spatial.search.CartesianShapeQueryTests.testQueryRandomGeoCollection" -Dtests.seed=D42035A03986D727 -Dtests.locale=lt -Dtests.timezone=Asia/Baku -Druntime.java=17`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.CartesianShapeQueryTests&tests.test=testQueryRandomGeoCollection

**Failure excerpt:**
```
org.elasticsearch.index.mapper.MapperParsingException: failed to parse field [geo] of type [shape]

  at __randomizedtesting.SeedInfo.seed([D42035A03986D727:CCF71E7B64748804]:0)
  at org.elasticsearch.index.mapper.AbstractGeometryFieldMapper.lambda$parse$1(AbstractGeometryFieldMapper.java:191)
  at org.elasticsearch.xpack.spatial.index.mapper.ShapeParser.parse(ShapeParser.java:40)
  at org.elasticsearch.index.mapper.AbstractGeometryFieldMapper.parse(AbstractGeometryFieldMapper.java:187)
  at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:409)
  at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:461)
  at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:313)
  at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:278)
  at org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:127)
  at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:79)
  at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:78)
  at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:999)
  at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:948)
  at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:892)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:321)
  at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:187)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:253)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:133)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:72)
  at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:211)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  at java.lang.Thread.run(Thread.java:833)

  Caused by: java.lang.IllegalArgumentException: at least three non-collinear points required

    at org.apache.lucene.geo.Tessellator.tessellate(Tessellator.java:174)
    at org.apache.lucene.geo.Tessellator.tessellate(Tessellator.java:152)
    at org.apache.lucene.document.XYShape.createIndexableFields(XYShape.java:83)
    at org.apache.lucene.document.XYShape.createIndexableFields(XYShape.java:72)
    at org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer$LuceneGeometryVisitor.visit(CartesianShapeIndexer.java:114)
    at org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer$LuceneGeometryVisitor.visit(CartesianShapeIndexer.java:50)
    at org.elasticsearch.geometry.Polygon.visit(Polygon.java:84)
    at org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer$LuceneGeometryVisitor.visit(CartesianShapeIndexer.java:66)
    at org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer$LuceneGeometryVisitor.visit(CartesianShapeIndexer.java:50)
    at org.elasticsearch.geometry.GeometryCollection.visit(GeometryCollection.java:51)
    at org.elasticsearch.xpack.spatial.index.mapper.CartesianShapeIndexer.indexShape(CartesianShapeIndexer.java:46)
    at org.elasticsearch.xpack.spatial.index.mapper.ShapeFieldMapper.index(ShapeFieldMapper.java:187)
    at org.elasticsearch.xpack.spatial.index.mapper.ShapeFieldMapper.index(ShapeFieldMapper.java:52)
    at org.elasticsearch.index.mapper.AbstractGeometryFieldMapper.lambda$parse$0(AbstractGeometryFieldMapper.java:187)
    at org.elasticsearch.xpack.spatial.index.mapper.ShapeParser.parse(ShapeParser.java:37)
    at org.elasticsearch.index.mapper.AbstractGeometryFieldMapper.parse(AbstractGeometryFieldMapper.java:187)
    at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:409)
    at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:461)
    at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:313)
    at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:278)
    at org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:127)
    at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:79)
    at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:78)
    at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:999)
    at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:948)
    at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:892)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:321)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:187)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:253)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:133)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:72)
    at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:211)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:833)

```

--- a/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/CartesianShapeQueryTestCase.java
+++ b/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/CartesianShapeQueryTestCase.java
@@ -7,13 +7,11 @@
 
 package org.elasticsearch.xpack.spatial.search;
 
-import org.apache.lucene.tests.geo.GeoTestUtil;
 import org.elasticsearch.common.geo.GeometryNormalizer;
 import org.elasticsearch.common.geo.Orientation;
 import org.elasticsearch.geometry.Geometry;
 import org.elasticsearch.geometry.GeometryCollection;
 import org.elasticsearch.geometry.Line;
-import org.elasticsearch.geometry.LinearRing;
 import org.elasticsearch.geometry.Point;
 import org.elasticsearch.geometry.Polygon;
 import org.elasticsearch.search.geo.BaseShapeQueryTestCase;
@@ -71,8 +69,7 @@ public abstract class CartesianShapeQueryTestCase extends BaseShapeQueryTestCase
     }
 
     protected Polygon nextPolygon() {
-        org.apache.lucene.geo.Polygon randomPoly = GeoTestUtil.nextPolygon();
-        return new Polygon(new LinearRing(randomPoly.getPolyLons(), randomPoly.getPolyLats()));
+        return ShapeTestUtils.randomPolygon(false);
     }
 
     protected Polygon nextPolygon2() {

--- a/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/CartesianShapeQueryTests.java
+++ b/x-pack/plugin/spatial/src/internalClusterTest/java/org/elasticsearch/xpack/spatial/search/CartesianShapeQueryTests.java
@@ -21,7 +21,6 @@ public class CartesianShapeQueryTests extends CartesianShapeQueryTestCase {
     }
 
     @Override
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/88682")
     public void testQueryRandomGeoCollection() throws Exception {
         super.testQueryRandomGeoCollection();
     }

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeTestUtils.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeTestUtils.java
@@ -27,9 +27,12 @@ import java.util.List;
 import java.util.function.Function;
 
 import static org.elasticsearch.geo.GeometryTestUtils.linearRing;
+import static org.elasticsearch.test.ESTestCase.randomValueOtherThanMany;
 
 /** generates random cartesian shapes */
 public class ShapeTestUtils {
+
+    public static final double MIN_VALID_AREA = 1e-10;
 
     public static double randomValue() {
         return XShapeTestUtil.nextDouble();
@@ -80,7 +83,7 @@ public class ShapeTestUtils {
     }
 
     public static Polygon randomPolygon(boolean hasAlt) {
-        XYPolygon lucenePolygon = XShapeTestUtil.nextPolygon();
+        XYPolygon lucenePolygon = randomValueOtherThanMany(p -> area(p) <= MIN_VALID_AREA, XShapeTestUtil::nextPolygon);
         if (lucenePolygon.numHoles() > 0) {
             XYPolygon[] luceneHoles = lucenePolygon.getHoles();
             List<LinearRing> holes = new ArrayList<>();
@@ -94,6 +97,16 @@ public class ShapeTestUtils {
             );
         }
         return new Polygon(linearRing(floatsToDoubles(lucenePolygon.getPolyX()), floatsToDoubles(lucenePolygon.getPolyY()), hasAlt));
+    }
+
+    static double area(XYPolygon p) {
+        double windingSum = 0;
+        final int numPts = p.numPoints() - 1;
+        for (int i = 0; i < numPts; i++) {
+            // compute signed area
+            windingSum += p.getPolyX(i) * p.getPolyY(i + 1) - p.getPolyY(i) * p.getPolyX(i + 1);
+        }
+        return Math.abs(windingSum / 2);
     }
 
     static double[] floatsToDoubles(float[] f) {

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeUtilTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeUtilTests.java
@@ -1,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.util;
+
+import org.apache.lucene.geo.XShapeTestUtil;
+import org.apache.lucene.geo.XYPolygon;
+import org.apache.lucene.geo.XYRectangle;
+import org.elasticsearch.test.ESTestCase;
+
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+
+public class ShapeUtilTests extends ESTestCase {
+    public void testBox() {
+        XYRectangle geom = XShapeTestUtil.nextBox();
+        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+    }
+
+    public void testPolygon() {
+        XYPolygon geom = XShapeTestUtil.nextPolygon();
+        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+        assertThat("Area should be non-zero", ShapeTestUtils.area(geom), greaterThan(0.0));
+    }
+
+    public void testFlatRectangle() {
+        XYPolygon geom = new XYPolygon(
+            new float[] { 54.69f, 54.69f, 180.0f, 180.0f, 54.69f },
+            new float[] { -2.80E-33f, 5.85E-33f, 5.85E-33f, -2.80E-33f, -2.80E-33f }
+        );
+        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+        assertThat(
+            "This flat rectangle has area less than allowed threshold",
+            ShapeTestUtils.area(geom),
+            lessThan(ShapeTestUtils.MIN_VALID_AREA)
+        );
+    }
+}

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeUtilTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/util/ShapeUtilTests.java
@@ -18,15 +18,15 @@ import static org.hamcrest.Matchers.lessThan;
 public class ShapeUtilTests extends ESTestCase {
     public void testBox() {
         XYRectangle geom = XShapeTestUtil.nextBox();
-        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
-        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+        assertThat("Geometry minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Geometry minY should be less than maxY", geom.minY, lessThan(geom.maxY));
     }
 
     public void testPolygon() {
         XYPolygon geom = XShapeTestUtil.nextPolygon();
-        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
-        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
-        assertThat("Area should be non-zero", ShapeTestUtils.area(geom), greaterThan(0.0));
+        assertThat("Geometry minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Geometry minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+        assertThat("Geometry area should be non-zero", ShapeTestUtils.area(geom), greaterThan(0.0));
     }
 
     public void testFlatRectangle() {
@@ -34,8 +34,8 @@ public class ShapeUtilTests extends ESTestCase {
             new float[] { 54.69f, 54.69f, 180.0f, 180.0f, 54.69f },
             new float[] { -2.80E-33f, 5.85E-33f, 5.85E-33f, -2.80E-33f, -2.80E-33f }
         );
-        assertThat("Box minX should be less than maxX", geom.minX, lessThan(geom.maxX));
-        assertThat("Box minY should be less than maxY", geom.minY, lessThan(geom.maxY));
+        assertThat("Geometry minX should be less than maxX", geom.minX, lessThan(geom.maxX));
+        assertThat("Geometry minY should be less than maxY", geom.minY, lessThan(geom.maxY));
         assertThat(
             "This flat rectangle has area less than allowed threshold",
             ShapeTestUtils.area(geom),

